design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+  * Support for reading & writing key schemas to & from the schema registry. 
+
+## What is not in scope
+
+ * Support for multiple key columns: this will come later.
+ * Support for single key columns _wrapped_ in an envelope of some kind: this will come later.
+ * Support for complex key column data types, i.e. array, struct and map 
+ * Support for `PROTOBUF` keys, as this requires support for wrapped keys: this will come later.
+
+## Value/Return
+
+We know from customers and community members that there are a lot of people that have data with 
+non-`KAFKA` formatted keys. This is the first step to unlocking that data and use-cases.
+
+With support for `AVRO` and `JSON` key formats there are a lot of existing use-cases that suddenly
+no longer require pre-processing, or tricky Connect SMTs configured and there are new use-cases,
+which ksqlDB was previously unsuitable for, as documented in the motivation section, which it can 
+now handle.
+
+## Public APIS
+
+### CREATE properties
+
+The following new properties will be accepted in the `WITH` clause of `CREATE` statements for streams
+and tables.
+
+* `KEY_FORMAT`: sets the key format, works long the same lines as the existing `VALUE_FORMAT`.
+* `FORMAT`: sets both the key and value format with a single property.
+
+The existing `VALUE_FORMAT` property will no longer be required, if the server config provides a 
+default. (See below).
+
+### Server configs
+
+The following new configuration options will be added. These configurations can be set globally, 
+within the application property file, or locally, via the `SET` command.
+
+* `ksql.persistence.default.format.key`: the default key format.
+* `ksql.persistence.default.format.value`: the default value format.
+
+## Design
+
+### New CREATE properties
+
+The new `KEY_FORMAT` or `FORMAT` property will be supported where ever the current `VALUE_FORMAT` is
+supported. Namely:
+
+ * In `CRREATE STREAM` and `CREATE TABLE` statements.
+ * In `CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements.
+ 
+Providing `FORMAT` will set both the key and value formats. Providing `FORMAT` along with either 
+`KEY_FORMAT` or `VALUE_FORMAT` will result in an error.
+
+The key format will follow the same inheritance rules as the current value format. Namely: any 
+derived stream will inherit the format of its leftmost source, unless the format is explicitly set
+in the with clause.
+
+For example:
+
+```sql
+-- Creates a table over a changelog topic with AVRO key and JSON value:
+CREATE TABLE USERS (
+    ID BIGINT PRIMARY KEY,
+    NAME STRING
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    KEY_FORMAT='AVRO',
+    VALUE_FORMAT='JSON'
+  );
+
+-- Creates a stream over a topic with JSON key and value:
+CREATE STREAM BIDS (
+    ITEM_ID BIGINT KEY,
+    USER_ID BIGINT,
+    AMMOUNT INT
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    FORMAT='JSON'
+  );
+
+-- Change the key format of a stream:
```

Comment:
When does the format change happen? Or this this unspecified by definition?

For example, the KS program could be `jsonStreamBids.selectKey(/*change the object from JSON to AVRO*)` (on read) or it could be `selectKey(/*change the object from JSON to AVRO*).to("AVRP_BIDS")` (on write).

For the simple `SELECT *` query from the example there is no difference, but assume you do anything stateful the data format of the state stores might differ -- also if you do joins, it might differ (and might require repartitioning).

Overall, it might be good to give the optimizer some freedom when to do the format change and thus it might be good to only guarantee that the _output_ stream will have the format but no guarantee about intermediate formats.

Reply:
Internal topics are... internal :).  

I'll add a section to the compatibility section to cover this.

Reply:
Add https://github.com/confluentinc/ksql/blob/dcd28ff1b26705625a477db4e3e81a1891b184aa/design-proposals/klip-33-key-format.md#internal-topics

In general, I'm saying we don't need to make any guarantees from release to release about the internal formats of topics. We're free to change them as we want.  

Of course, we _do_ maintain the internal formats for _existing_ queries. But that's handled already by capturing the internal format in the query plan.